src/turbine/turbine_tree.zig

```diff
@@ -0,0 +1,804 @@
+        // nodes with a stake of 0
+        var result = try std.ArrayList(Node).initCapacity(allocator, nodes.items.len);
+        errdefer result.deinit();
+        var ip_counts = std.AutoArrayHashMap(IpAddr, usize).init(allocator);
+        defer ip_counts.deinit();
+        for (nodes.items) |node| {
+            // Add the node to the result if it does not exceed the
+            // maximum number of nodes per IP
+            var node_appended = false;
+            if (node.tvuAddress()) |tvu_addr| {
+                const ip_count = ip_counts.get(tvu_addr.ip()) orelse 0;
+                if (ip_count < MAX_NODES_PER_IP_ADDRESS) {
+                    node_appended = true;
+                    result.appendAssumeCapacity(node);
+                }
+                try ip_counts.put(tvu_addr.ip(), ip_count + 1);
+            }
+
+            // Keep the node for deterministic shuffle but remove
+            // contact info so that it is not used for retransmit
+            if (!node_appended and node.stake > 0) {
+                result.appendAssumeCapacity(.{
+                    .id = .{ .pubkey = node.pubkey() },
+                    .stake = node.stake,
+                });
+            }
+        }
+
+        return result;
+    }
+
+    /// Get Tvu peers from the gossip table, that is all peers with a matching
+    /// shred version.
+    fn getTvuPeers(
+        allocator: std.mem.Allocator,
+        my_contact_info: ThreadSafeContactInfo,
+        gossip_table_rw: *RwMux(GossipTable),
+    ) !std.ArrayList(ThreadSafeContactInfo) {
+        const gossip_table, var gossip_table_lg = gossip_table_rw.readWithLock();
+        defer gossip_table_lg.unlock();
+
+        var contact_info_iter = gossip_table.contactInfoIterator(0);
+        var tvu_peers = try std.ArrayList(ThreadSafeContactInfo).initCapacity(allocator, gossip_table.contact_infos.count());
+
+        while (contact_info_iter.nextThreadSafe()) |contact_info| {
+            if (!contact_info.pubkey.equals(&my_contact_info.pubkey) and contact_info.shred_version == my_contact_info.shred_version) {
+                tvu_peers.appendAssumeCapacity(contact_info);
+            }
+        }
+
+        return tvu_peers;
+    }
+};
+
+/// TestEnvironment sets up the dependencies for testing the TurbineTree.
+/// Testing could be made more thorough by constructing the gossip table and
+/// bank fields, and using the TurbineTreeProvider but this is sufficient for now.
+const TestEnvironment = struct {
+    allocator: std.mem.Allocator,
+    my_contact_info: ThreadSafeContactInfo,
+    gossip_table_tp: ThreadPool,
+    gossip_table_rw: RwMux(GossipTable),
+    staked_nodes: std.AutoArrayHashMap(Pubkey, u64),
+
+    pub fn init(
+        allocator: std.mem.Allocator,
+        rng: std.rand.Random,
+        num_known_nodes: usize,
+        num_unknown_staked_nodes: usize,
+        known_nodes_unstaked_ratio: struct { u64, u64 },
```

Comment:
Named parameter struct for at least these three parameters. The four numbers in `TestEnvironment.init(allocator, rng, 1_000, 100, .{ 1, 7 })` feel pretty magical.